doc/source/train/user-guides/fault-tolerance.rst

```diff
@@ -5,128 +5,198 @@
-passed to the ``Trainer``:
+passed to the ``Trainer``. By default, worker fault tolerance is disabled with ``max_failures=0``.
 
-.. literalinclude:: ../doc_code/key_concepts.py
+.. literalinclude:: ../doc_code/fault_tolerance.py
     :language: python
     :start-after: __failure_config_start__
     :end-before: __failure_config_end__
 
+Altogether, this is what an example Torch training script with worker fault tolerance looks like:
+
+.. literalinclude:: ../doc_code/fault_tolerance.py
+    :language: python
+    :start-after: __worker_fault_tolerance_start__
+    :end-before: __worker_fault_tolerance_end__
+
+
 Which checkpoint will be restored?
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-Ray Train will automatically resume training from the latest available
+
+Ray Train will populate :func:`ray.train.get_checkpoint() <ray.train.get_checkpoint>` with the latest available
 :ref:`checkpoint reported to Ray Train <train-checkpointing>`.
+The :class:`~ray.train.Checkpoint` object returned by this method has the
+:meth:`~ray.train.Checkpoint.as_directory` and :meth:`~ray.train.Checkpoint.to_directory` methods
+to download the checkpoint from the :class:`RunConfig(storage_path) <ray.train.RunConfig>` to local disk.
+
+.. note::
+    :meth:`~ray.train.Checkpoint.as_directory` and :meth:`~ray.train.Checkpoint.to_directory`
+    will only download the checkpoint once per node even if there are multiple workers on the node.
+    The workers share the same checkpoint directory on local disk.
+
+Illustrated Example
+~~~~~~~~~~~~~~~~~~~
+
+Consider the following example of a cluster containing a CPU head node and 2 GPU worker nodes.
+There are 4 GPU training workers running on the 2 worker nodes.
+The :ref:`storage path has been configured <persistent-storage-guide>` to use cloud storage, which is where checkpoints are saved.
+
+.. figure:: ../images/fault_tolerance/worker_node_failure.png
+    :align: left
+
+    (Left) Training has been running for some time, and the latest checkpoint has been saved to cloud storage.
+
+    (Right) One of the worker GPU nodes fails due to a hardware fault. Ray Train detects this failure and shuts down all the workers.
+    Since the number of failures detected so far is less than the configured ``max_failures``, Ray Train will attempt to restart training,
+    rather than exiting and raising an error.
+
+.. figure:: ../images/fault_tolerance/worker_node_replacement.png
+    :align: left
+
+    (Left) Same as the previous figure.
+
+    (Right) Ray Train has requested a new worker node to join the cluster and is waiting for it to come up.
+
+.. figure:: ../images/fault_tolerance/resume_from_checkpoint.png
+    :align: left
+
+    (Left) Same as the previous figure.
+
+    (Right) The new worker node has joined the cluster.
+    Ray Train restarts all the worker processes and provides them with the latest checkpoint.
+    The workers download the checkpoint from storage and use it to resume training.
```

Comment:
What's the reason for copying the image with `(Left) Same as the previous figure.`? Could we just remove all copies? So in this case we would have 4 figures total, no copies.

I also did notice that it's hard to read the text of the image on my laptop, so removing side-by-side figures might help zoom it in.

Reply:
sounds good

Reply:
the reason was to showcase the "transitions", but can also do that vertically